src/table/builder.rs

```diff
@@ -114,36 +136,45 @@ impl Builder {
     }
 
     pub fn reach_capacity(&self, capacity: u64) -> bool {
-        let block_size = self.buf.len() + self.entry_offsets.len() * 4 + 4 + 8 + 4;
-        let estimated_size = block_size + 4 + 5 * self.table_index.offsets.len();
-        estimated_size > block_size
+        let block_size = self.buf.len() as u32 + // length of buffer
+                                 self.entry_offsets.len() as u32 * 4 + // all entry offsets size
+                                 4 + // count of all entry offsets
+                                 8 + // checksum bytes
+                                 4; // checksum length
+        let estimated_size = block_size + 
+                                  4 + // index length
+                                  5 * self.table_index.offsets.len() as u32; // TODO: why 5?
+        estimated_size as u64 > capacity
     }
 
     pub fn finish(&mut self) -> Bytes {
         self.finish_block();
-        let bytes = self.table_index.write_to_bytes().unwrap();
+        if self.buf.is_empty() {
+            return Bytes::new();
+        }
+        let mut bytes = BytesMut::new();
+        self.table_index.encode(&mut bytes).unwrap();
         assert!(bytes.len() < u32::MAX as usize);
         self.buf.put_slice(&bytes);
-        self.buf.put_u32_le(bytes.len());
+        self.buf.put_u32_le(bytes.len() as u32);
         let cs = self.build_checksum(&bytes);
         self.write_checksum(cs);
-        self.buf.freeze()
+        self.buf.clone().freeze()
```

Comment:
This freeze may introduce overhead. I'll investigate how to fix this later.

Reply:
Freezing a `BytesMut` requires full ownership of `buf`. But in `finish(&mut self)`, we could only get a `&mut buf`. I'm not sure whether calling `finish` allows us to give away `builder`. I'll change this to `finish(self)` and remove `clone()` later, if `builder` won't be used after `finish`.